File: scripts/statfor.py
#!/usr/bin/env python3
"""Read a stream of data from a file, perform a statistical analysis"""
import sys
import math
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab


MAX_STEPS = 200


def int_width(num):
    """Return decimal with of integer number"""
    return math.floor(math.log10(num))

def read_data(fpath):
    """Read data from file located at fpath"""
    return np.array(list(float(line.split()[0]) for line in open(fpath, "r")))


def correlation(data, mean, var):
    """Calculate correlation time, effective N, error"""
    nsteps = len(data)
    tcorr = 1
    # needed for formatting
    max_i = min(MAX_STEPS, nsteps -1)
    max_width = int_width(max_i)
    f = 1
    with open("corr.out", "w") as file:
        for i in range(1, max_i + 1):
            corr = ((data[:nsteps-i] - mean)*(data[i:] - mean))[:nsteps-i].sum()
            corr /= (nsteps-i)*var
            if corr < 0:
                f = 0
            tcorr += 2*corr*f
            file.write(f"{i}" +" "*(max_width-int_width(i)+4) + f"{corr:.10e}\n")
        tcorr = max(1, tcorr)

    return tcorr, nsteps/tcorr, math.sqrt(var*tcorr/nsteps)


def blocking(data):
    """Perform blocking analysis"""
    ndata = len(data)
    min_left = 100
    nsizes = 100
    large = int(ndata/min_left)
    step_size = int(max(1, large/nsizes))

    # needed for formatting
    max_width = int_width(large+1)
    
    sizes, errors = [], [] 
    with open("blocking.out", "w") as file:
        for size in range(1, large+1, step_size):
            nblocks = int(ndata/size)
            # block averages
            blocks = np.array_split(data, nblocks)
            averages_blk = np.array(list(map(np.mean, blocks)))
            averages_sq_blk = np.array([ab**2 for ab in averages_blk]).mean()
            # estimated error of mean at this block size
            errors.append((np.sqrt((averages_sq_blk - averages_blk.mean()**2)\
                    /(nblocks-1))))
            sizes.append(size)
            file.write(f"{size}" \
                    +" "*(max_width-int_width(size)+4) + f"{errors[-1]:.10e}\n")

    return np.array(errors), np.array(sizes)


def histogram(data, bins=12):
    hist, edges = np.histogram(data, bins)
    with open("histo.out", "w") as file:
        for edge, h in zip(edges, hist):
            file.write(f"{edge:.10e} {h:.10e}\n")
    return hist, edges



if __name__ == "__main__":
    SERIES = read_data(sys.argv[1])

    MEAN = SERIES.mean()
    VAR = SERIES.var(ddof=1)

    TCORR, NEFF, SIGMA = correlation(SERIES, MEAN, VAR)

    ERRORS, SIZES = blocking(SERIES)

    hist, bin_edges = histogram(SERIES)
    
    if "-p" in sys.argv:
        fig = plt.figure()
        ax = [fig.add_subplot(121), fig.add_subplot(122)]

        ax[0].plot(SIZES, ERRORS, '.')
        ax[0].set_ylabel("Error")
        ax[0].set_xlabel("Block size")
        ax[0].grid()
        
        blksize = 1
        blocks = np.array_split(SERIES, len(SERIES)//blksize)
        means = np.array([b.mean() for b in blocks])

        n, bins = ax[1].hist(means, density=True, bins=len(means)//10)[:2]
        gaussian = mlab.normpdf(bins, means.mean(), np.sqrt(means.var()))
        ax[1].plot(bins, gaussian)
        ax[1].grid()
        ax[1].set_title("Block average histogram")

        plt.show()

    # Print output
    print(f"average\t\t{MEAN:.10f}")
    print(f"variance\t{VAR:.10f}")
    print(f"t corr\t\t{TCORR:.10f}")
    print(f"n eff\t\t{NEFF:.10f}")
    print(f"sigma\t\t{SIGMA:.10f}")

File: scripts/statfor.rs
use std::vec::Vec;
use std::io::{self, BufRead, Write, Result};
use std::fs::File;


/// Read a stream of data from stdin line by line
fn read_data() -> Result<Vec<f64>> {
    let stdin = io::stdin();
    let data = stdin.lock().lines()
        .map(|x| x.unwrap().split_whitespace().collect::<Vec<_>>()[1]
             .parse::<f64>().expect("Failed to parse as float"))
        .collect::<Vec<_>>();
    Ok(data)
}


/// Compute the mean of an array of floats
fn mean(data: &[f64]) -> f64 {
    data.iter().sum::<f64>()/(data.len() as f64)
}


/// Compute the variance of an array of floats, with given DOF
fn variance(data: &[f64], ddof: usize) -> f64 {
    let average = mean(data);
    data.iter().map(|x| (average - x).powi(2)).sum::<f64>()/(data.len() -ddof) as f64
}


/// Compute the correlation time, effective steps, and sigma of an array of data
fn correlation(data: &[f64], mean: f64, variance: f64) -> Result<(f64, f64, f64)> {
    const MAX_STEPS: usize = 200;
    let nsteps = data.len();
    let mut tcorr: f64 = 1.0;
    let max_i = MAX_STEPS.min(nsteps - 1);

    let mut file = File::create("corr.out")?;
    let mut f = 1.0;

    for i in 1..=max_i {
        let mut corr = data[..nsteps-i].iter().zip(data[i..].iter())
            .map(|(&x, &xshift)| (x - mean)*(xshift - mean))
            .take(nsteps - i)
            .sum::<f64>();
        corr /= variance*(nsteps - i) as f64;
        if corr < 0. {
            f = 0.0;
        }
        tcorr += 2.0 * corr * f;
        file.write(format!("{} {}\n", i, corr).as_bytes())?;
    }
    tcorr = tcorr.max(1.0);
    Ok((tcorr, nsteps as f64 / tcorr, (variance*tcorr / nsteps as f64).sqrt()))
}


/// Perform blocking analysis over data
fn blocking(data: &[f64]) -> Result<()> {
    let ndata = data.len();
    const MIN_LEFT: usize = 20;
    const NSIZES: usize = 100;
    let large = ndata / MIN_LEFT;
    let step_size = (large/NSIZES).max(1);
    
    let mut file = File::create("blocking.out")?;

    for size in (1..=large).step_by(step_size) {
        let nblocks = ndata/size;
        // block averages
        let blocks = data.chunks(size);
        let ave_blks = blocks.map(mean);
        let ave_sq = mean(&ave_blks.clone().map(|x| x.powi(2)).collect::<Vec<_>>());
        let ave = mean(&ave_blks.collect::<Vec<_>>());
        // estimated error at this block size
        let error = ((ave_sq - ave.powi(2))/(nblocks - 1) as f64).sqrt();
        // write to file
        file.write(format!("{} {}\n", size, error).as_bytes())?;
    }

    Ok(())
}


fn main() {
    let data = read_data().expect("Failed to parse input stream");
    let data = &data[..data.len()-1];
    let average = mean(&data);
    let var = variance(&data, 1);
    let (tcorr, neff, sigma) = correlation(&data, average, var)
        .expect("Failed to write correlation to file");

    blocking(&data).expect("Failed to write blocking data to file");

    println!("average\t\t{}", average);
    println!("variance\t{}", var);
    println!("tcorr\t\t{}", tcorr);
    println!("n eff\t\t{}", neff);
    println!("sigma\t\t{}", sigma);
}
